mirrord/agent/src/file.rs

```diff
@@ -362,34 +367,37 @@ impl FileManager {
             })
     }
 
-    #[tracing::instrument(level = "trace", skip(self))]
+   #[tracing::instrument(level = "trace", skip(self))]
     pub(crate) fn read_limited(
-        &mut self,
-        fd: u64,
-        buffer_size: u64,
-        start_from: u64,
+    &mut self,
+    fd: u64,
+    buffer_size: u64,
+    start_from: u64,
     ) -> RemoteResult<ReadFileResponse> {
-        self.open_files
-            .get_mut(&fd)
-            .ok_or(ResponseError::NotFound(fd))
-            .and_then(|remote_file| {
-                if let RemoteFile::File(file) = remote_file {
-                    let mut buffer = vec![0; buffer_size as usize];
-
-                    let read_result = file.read_at(&mut buffer, start_from).map(|read_amount| {
-                        // We handle the extra bytes in the `pread` hook, so here we can just
-                        // return the full buffer.
-                        ReadFileResponse {
-                            bytes: buffer,
-                            read_amount: read_amount as u64,
-                        }
-                    })?;
-
-                    Ok(read_result)
-                } else {
-                    Err(ResponseError::NotFile(fd))
-                }
-            })
+    self.open_files
+        .get_mut(&fd)
+        .ok_or(ResponseError::NotFound(fd))
+        .and_then(|remote_file| {
+            if let RemoteFile::File(file) = remote_file {
+                let mut buffer = vec![0; buffer_size as usize];
+
+                let read_amount = file.read_at(&mut buffer, start_from)?;
+
+                // Further optimization: Slice the buffer based on the actual number of bytes read
+                let read_bytes = buffer[..read_amount].to_vec();
```

Comment:
`.to_vec` is copying buffer slice into another vector, similar desired result can be achieved with `.truncate(read_amount)` where the `buffer` Vec should shrink to expected size (same for read function)

Reply:
Will fix, thank you for feedback!